kaldb/src/main/java/com/slack/kaldb/logstore/SchemaAwareLogDocumentBuilderImpl.java

```diff
@@ -0,0 +1,553 @@
+        if (fieldDef.storeNumericDocValue) {
+          doc.add(new NumericDocValuesField(name, value));
+        }
+      }
+    },
+    FLOAT("float") {
+      @Override
+      public void addField(Document doc, String name, Object v, FieldDef fieldDef) {
+        float value = (float) v;
+        if (fieldDef.isIndexed) {
+          doc.add(new FloatPoint(name, value));
+        }
+        if (fieldDef.isStored) {
+          doc.add(new StoredField(name, value));
+        }
+        if (fieldDef.storeNumericDocValue) {
+          doc.add(new FloatDocValuesField(name, value));
+        }
+      }
+    },
+    DOUBLE("double") {
+      @Override
+      public void addField(Document doc, String name, Object v, FieldDef fieldDef) {
+        double value = (double) v;
+        if (fieldDef.isIndexed) {
+          doc.add(new DoublePoint(name, value));
+        }
+        if (fieldDef.isStored) {
+          doc.add(new StoredField(name, value));
+        }
+        if (fieldDef.storeNumericDocValue) {
+          doc.add(new DoubleDocValuesField(name, value));
+        }
+      }
+    },
+    BOOLEAN("boolean") {
+      @Override
+      public void addField(Document doc, String name, Object value, FieldDef fieldDef) {
+        // Lucene has no native support for Booleans so store that field as text.
+        if ((boolean) value) {
+          addTextField(doc, name, "true", fieldDef);
+        } else {
+          addTextField(doc, name, "false", fieldDef);
+        }
+      }
+    };
+
+    private final String name;
+
+    FieldType(String name) {
+      this.name = name;
+    }
+
+    public abstract void addField(Document doc, String name, Object value, FieldDef fieldDef);
+
+    public String getName() {
+      return name;
+    }
+
+    static Object convertFieldValue(Object value, FieldType fromType, FieldType toType) {
+      // String type
+      if (fromType == FieldType.TEXT) {
+        if (toType == FieldType.INTEGER) {
+          return Integer.valueOf((String) value);
+        }
+        if (toType == FieldType.LONG) {
+          return Long.valueOf((String) value);
+        }
+        if (toType == FieldType.FLOAT || toType == FieldType.DOUBLE) {
+          return Double.valueOf((String) value);
+        }
+      }
+
+      // Int type
+      if (fromType == FieldType.INTEGER) {
+        if (toType == FieldType.TEXT) {
+          return ((Integer) value).toString();
+        }
+        if (toType == FieldType.LONG) {
+          return ((Integer) value).longValue();
+        }
+        if (toType == FieldType.FLOAT) {
+          return ((Integer) value).floatValue();
+        }
+        if (toType == FieldType.DOUBLE) {
+          return ((Integer) value).doubleValue();
+        }
+      }
+
+      // Long type
+      if (fromType == FieldType.LONG) {
+        if (toType == FieldType.TEXT) {
+          return ((Long) value).toString();
+        }
+        if (toType == FieldType.INTEGER) {
+          return ((Long) value).intValue();
```

Comment:
I'm still looking over this and not sure I have all of the context, yet, but a couple comments as I read through it. It seems like this would silently truncate data in the case of narrowing conversions such as `long` -> `int` where the value exceeds `Integer.MAX_VALUE`, `double` -> `float`, etc. Also, I didn't see handling of conversions that would result in runtime exceptions such as `NumberFormatException` if there was an attempt to convert an invalid string to a numeric value.

Reply:
Loss of precision is expected. Handling exceptions is a good one. I will add more unit tests for it.

Reply:
It's not just a loss of precision but arithmetic overflow and such where the converted value is meaningless. I expect there will have to be some kind of handling for fields that cannot be converted to the target type when the `FieldConflictPolicy` is `CONVERT_FIELD_VALUE` or `CONVERT_AND_DUPLICATE_FIELD`, but that seems like something that could be worked out later.

Reply:
Good callout on handling exceptions. Fixed up the code to handle exceptions and added more tests. Also, added tests to account for loss in precision and not errors. The converted values are meaningless, but leaving them as is for now. We can add an additional tag like converted fields etc.. to the document in a future iteration.